src/globus_cli/parsing/param_types/json_strorfile.py

```diff
@@ -0,0 +1,107 @@
+from __future__ import annotations
+
+import dataclasses
+import json
+import os
+import sys
+import typing as t
+
+import click
+
+from globus_cli.constants import EXPLICIT_NULL, ExplicitNullType
+from globus_cli.types import JsonValue
+
+from .annotated_param import AnnotatedParamType
+
+
+@dataclasses.dataclass
+class ParsedJSONData:
+    # str means it's the path to a file input
+    # including `-` for stdin
+    # None means the parse came from an argument input
+    filename: t.Union[str, None]
+    # data is the parsed data
+    data: JsonValue
+
+
+class JSONStringOrFile(AnnotatedParamType):
+    """
+    Parse an input which could be a filename or could be a JSON blob being
+    supplied on the commandline.
+
+    For example,
+        --opt './foo/bar.json'
+        --opt "$(./myscript.sh)"
+
+    For compatibility with the older mode of use, this param also supports a
+    `file:` prefix which disambiguates an input as being a filename. e.g.
+        --opt 'file:./foo/bar.json'
+
+    The parsing happens in the following order of precedence:
+        - 'file:' prefix (filename)
+        - bare filename, including `-` for stdin
+        - JSON parse
+
+    This ordering protects against odd behavior if a filename happens to be valid JSON.
+    Most likely to happen with a JSON literal like an integer, e.g. `--opt 100`
+
+    The output of the parse is a loaded JSON object wrapped in a dataclass which
+    contains contextual information. This can be useful for providing better error
+    messages.
+    """
+
+    def __init__(
+        self, *args: t.Any, null: t.Any | None = None, **kwargs: t.Any
+    ) -> None:
+        self.null = null
+        super().__init__(*args, **kwargs)
+
+    def get_metavar(self, param: click.Parameter) -> str:
+        return "[JSON_FILE|JSON|file:JSON_FILE]"
```

Comment:
My gut is that this won't be completely clear to a user. Does this change seem clearer, or is it already clear enough?

```suggestion
        return "[JSON_FILE|INLINE_JSON|file:JSON_FILE]"
```

Reply:
I don't know that anything will suffice to totally clarify for users other than an example usage.

I'd be somewhat inclined to leave it as-is? I worry that `INLINE_JSON` could generate more confusion than the bare `JSON` -- e.g. users thinking that it's some limitation on the format to have no newlines.

This is all weakly held, but that's where I'm tending.

Reply:
Makes sense. I'm on board!